src/App.tsx

```diff
@@ -1,25 +1,62 @@
-import React from 'react';
+import { useEffect, useMemo, useState } from 'react';
+import { useLocation, useNavigate } from 'react-router-dom';
 import './App.css';
 import { getNumbers } from './utils';
+import { Pagination } from './components/Pagination';
 
 // eslint-disable-next-line @typescript-eslint/no-unused-vars
-const items = getNumbers(1, 42).map(n => `Item ${n}`);
 
 export const App: React.FC = () => {
+  const location = useLocation();
+  const navigate = useNavigate();
+
+  const queryParams = new URLSearchParams(location.search);
+  const initialPage = parseInt(queryParams.get('page') || '1', 10);
+  const initialPerPage = parseInt(queryParams.get('perPage') || '5', 10);
+
+  const [currentPage, setCurrentPage] = useState(initialPage);
+  const [perPage, setPerPage] = useState(initialPerPage);
+
+  const items = useMemo(() => getNumbers(1, 42).map(n => `Item ${n}`), []);
+
+  useEffect(() => {
+    const params = new URLSearchParams();
+
+    params.set('page', currentPage.toString());
+    params.set('perPage', perPage.toString());
+    navigate({ search: `?${params.toString()}` });
```

Comment:
just use `a` tag with the correct href as it was initially
So this entire useEffect is redundant